tests/llmcompressor/modifiers/quantization/test_shared_embeddings.py

```diff
@@ -0,0 +1,190 @@
+)
+from tests.testing_utils import requires_gpu
+
+
+@requires_gpu
+@pytest.mark.skipif(
+    (not os.getenv("HF_TOKEN")),
+    reason="Skipping correctness tests requiring gated model access",
+)
+def test_quantization_with_automatic_untie():
+    """
+    Test that quantization with automatic untie_word_embeddings produces the same
+    results as manually calling untie_word_embeddings first.
+
+    This test verifies the functionality added to QuantizationMixin where
+    _untie_if_target_shared is called during start_calibration to automatically
+    handle shared input/output embeddings when they are targeted for quantization.
+    """
+    model_id = "meta-llama/Llama-3.2-1B-Instruct"
+    scheme = "W8A8"
+    targets = ["Linear"]  # This targets lm_head (Linear layer)
+
+    # Test 1: Apply quantization WITHOUT manually untieing first
+    # (relies on automatic untieing in start_calibration)
+    model_auto = AutoModelForCausalLM.from_pretrained(
+        model_id, device_map="cuda", torch_dtype=torch.float32
+    )
+
+    # Verify embeddings are initially tied
+    input_embed = model_auto.get_input_embeddings()
+    output_embed = model_auto.get_output_embeddings()
+    initial_tied = input_embed.weight is output_embed.weight
+
+    state_auto = State(model=model_auto)
+    modifier_auto = QuantizationModifier(scheme=scheme, targets=targets)
+
+    input_data = {k: v.to("cuda") for k, v in model_auto.dummy_inputs.items()}
+
+    with torch.no_grad():
+        baseline_output = model_auto(**input_data)
+
+    # Initialize and start calibration (should automatically untie if needed)
+    modifier_auto.on_initialize(state_auto)
+    modifier_auto.on_start(state_auto, None)
+
+    with torch.no_grad():
+        output_auto = model_auto(**input_data)
+
+    # Verify that embeddings were untied
+    auto_untied = input_embed.weight is not output_embed.weight
+
+    # Clean up
+    del model_auto, state_auto, modifier_auto
+    torch.cuda.empty_cache()
+
+    # Test 2: Apply quantization WITH manual untieing first
+    model_manual = AutoModelForCausalLM.from_pretrained(
+        model_id, device_map="cuda", torch_dtype=torch.float32
+    )
+
+    # Manually untie before quantization
+    untie_word_embeddings(model_manual)
+
+    state_manual = State(model=model_manual)
+    modifier_manual = QuantizationModifier(scheme=scheme, targets=targets)
+
+    # Initialize and start calibration
+    modifier_manual.on_initialize(state_manual)
+    modifier_manual.on_start(state_manual, None)
+
+    with torch.no_grad():
+        output_manual = model_manual(**input_data)
+
+    # Verify that automatic and manual untieing produce identical results
+    # The outputs should be identical (or very close due to numerical precision)
+    mse = torch.nn.MSELoss()(output_auto.logits, output_manual.logits)
+
+    # For quantization, we expect very low MSE between the two approaches
+    # (they should be essentially identical)
+    assert mse < 1e-6, (
```

Comment:
If you feel comfortable, consider removing the MSE tests and only verifying that the embeddings are now untied / can be modified independently.

Reply:
will do